File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxHeight(TreeNode*root,int& diameter){
        if(root==nullptr){
            return 0;
        }
        int lh = maxHeight(root->left,diameter);
        int rh = maxHeight(root->right,diameter);
        diameter = max(diameter,lh+rh);
        return max(lh,rh)+1;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        int diameter = 0;
        maxHeight(root,diameter);
        return diameter;
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public static int diameter;
    public static int find_height(TreeNode root){
        if(root == null){
            return 0;
        }
        int left_height = find_height(root.left);
        int right_height = find_height(root.right);
        diameter = Math.max(diameter,left_height+right_height);

        return Math.max(left_height,right_height)+1;

    }
    public int diameterOfBinaryTree(TreeNode root) {
        diameter = 0;
        find_height(root);
        return diameter;
    }
}